crates/ruff_python_formatter/src/expression/expr_tuple.rs

```diff
@@ -1,19 +1,156 @@
 use crate::expression::parentheses::{
     default_expression_needs_parentheses, NeedsParentheses, Parentheses, Parenthesize,
 };
-use crate::{not_yet_implemented_custom_text, FormatNodeRule, PyFormatter};
-use ruff_formatter::{write, Buffer, FormatResult};
+use crate::trivia::find_first_non_trivia_character_after;
+use crate::{AsFormat, FormatNodeRule, PyFormatter, USE_MAGIC_TRAILING_COMMA};
+use ruff_formatter::formatter::Formatter;
+use ruff_formatter::prelude::{
+    block_indent, group, hard_line_break, if_group_breaks, soft_block_indent, soft_line_break,
+    soft_line_break_or_space, text,
+};
+use ruff_formatter::{format_args, write, Buffer, Format, FormatResult};
+use ruff_python_ast::prelude::{Expr, Ranged};
+use ruff_text_size::{TextLen, TextRange};
 use rustpython_parser::ast::ExprTuple;
 
 #[derive(Default)]
 pub struct FormatExprTuple;
 
 impl FormatNodeRule<ExprTuple> for FormatExprTuple {
-    fn fmt_fields(&self, _item: &ExprTuple, f: &mut PyFormatter) -> FormatResult<()> {
-        write!(f, [not_yet_implemented_custom_text("(1, 2)")])
+    fn fmt_fields(&self, item: &ExprTuple, f: &mut PyFormatter) -> FormatResult<()> {
+        let ExprTuple {
+            range,
+            elts,
+            ctx: _,
+        } = item;
+
+        // Handle the edge cases of an empty tuple and a tuple with one element
+        let last = match &elts[..] {
+            [] => {
+                return text("()").fmt(f);
+            }
+            [single] => {
+                return write!(
+                    f,
+                    [group(&format_args![
+                        // A single element tuple always needs parentheses
+                        &text("("),
+                        soft_block_indent(&group(&format_args![single.format(), &text(",")])),
+                        &text(")"),
+                    ])]
+                );
+            }
+            [.., last] => last,
+        };
+
+        let saved_level = f.context().node_level();
+        // Tell the children they need parentheses
+        f.context_mut().set_node_level(NodeLevel::Expression);
+
+        let magic_trailing_comma = USE_MAGIC_TRAILING_COMMA
+            && matches!(
+                find_first_non_trivia_character_after(last.range().end(), f.context().contents()),
+                Some((_, ','))
+            );
+
+        if magic_trailing_comma {
+            // A magic trailing comma forces us to print in expanded mode since we have more than
+            // one element
+            write!(
+                f,
+                [group(&format_args![
+                    // An expanded group always needs parentheses
+                    &text("("),
+                    hard_line_break(),
+                    block_indent(&ExprSequence::new(elts)),
+                    hard_line_break(),
+                    &text(")"),
+                ])]
+            )?;
+        } else if is_parenthesized(*range, elts, f) {
```

Comment:
We don't need to solve this here but we probably need to preserve the right amount of parentheses. 

Isn't the `Expr` parentheses handling already covering this case?

Reply:
yep, also didn't realize that this is now handled

Reply:
on second test, surprisingly not:

```
((1))
# item.range() = 2..3
```

```
((1,))
# item.range() = 1..5
# item.range() = 2..3
```

Reply:
Ah that makes sense. The expr handling doesn't preserve tuple parentheses. Can we end up with too many parentheses if the expr and tuple ends up adding them (maybe return Never in needs_parentheses when they are optional)

Reply:
👋 looking here after trying to implement `StmtFor`. for `for` loops, it seems black removes existing parens around the target `for (x, y) in z` -> `for x, y in z`

see e.g. https://github.com/astral-sh/ruff/blob/be107dad64510e3e34fdd322ba333da06b61be6c/crates/ruff_python_formatter/resources/test/fixtures/black/simple_cases/remove_for_brackets.py#L1

i'm still trying to learn by way around this codebase but i guess this needs to check the parenthesize options or something?